File: src/algorithms/dynamic_programming.cpp
#include "dynamic_programming.h"
#include <algorithm>
#include <vector>
#include <iostream>

/**
 * @brief Solve weighted activity selection using dynamic programming
 * 
 * This implements the classic weighted activity selection problem using
 * dynamic programming. Activities are first sorted by end time, then
 * we compute the optimal weight for each prefix.
 * 
 * @param activities Vector of activities with weights
 * @return Optimal schedule with maximum total weight
 * @complexity Time: O(n log n), Space: O(n)
 */
std::vector<Activity> DynamicProgrammingScheduler::solveWeightedActivitySelection(std::vector<Activity> activities) {
    if (activities.empty()) {
        return {};
    }
    
    int n = activities.size();
    
    // Sort activities by end time
    std::sort(activities.begin(), activities.end(), 
              [](const Activity& a, const Activity& b) {
                  return a.end < b.end;
              });
    
    // dp[i] = maximum weight achievable using activities 0..i
    std::vector<double> dp(n);
    std::vector<int> parent(n, -1); // To reconstruct solution
    
    // Base case
    dp[0] = activities[0].weight;
    
    // Fill DP table
    for (int i = 1; i < n; i++) {
        // Option 1: Don't include current activity
        double withoutCurrent = dp[i - 1];
        
        // Option 2: Include current activity
        double withCurrent = activities[i].weight;
        int latest = findLatestNonConflicting(activities, i);
        
        if (latest != -1) {
            withCurrent += dp[latest];
        }
        
        // Choose the better option
        if (withCurrent > withoutCurrent) {
            dp[i] = withCurrent;
            parent[i] = latest;
        } else {
            dp[i] = withoutCurrent;
            parent[i] = i - 1; // Mark as "don't include"
        }
    }
    
    // Reconstruct the optimal solution
    std::vector<Activity> result;
    int current = n - 1;
    
    while (current >= 0) {
        if (current == 0 || dp[current] != dp[current - 1]) {
            // This activity is included in the optimal solution
            result.push_back(activities[current]);
            current = parent[current];
        } else {
            // This activity is not included
            current--;
        }
    }
    
    // Reverse to get correct order
    std::reverse(result.begin(), result.end());
    
    return result;
}

/**
 * @brief Find the latest non-conflicting activity
 * 
 * Uses binary search to find the latest activity that ends before
 * the current activity starts.
 * 
 * @param activities Sorted activities by end time
 * @param index Current activity index
 * @return Index of latest non-conflicting activity, -1 if none
 */
int DynamicProgrammingScheduler::findLatestNonConflicting(const std::vector<Activity>& activities, int index) {
    int start = activities[index].start;
    
    // Binary search for the latest activity that ends before current starts
    int left = 0, right = index - 1;
    int result = -1;
    
    while (left <= right) {
        int mid = left + (right - left) / 2;
        
        if (activities[mid].end <= start) {
            result = mid;
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    
    return result;
}

/**
 * @brief Check if two activities have a time conflict
 * 
 * @param a1 First activity
 * @param a2 Second activity
 * @return true if activities overlap in time, false otherwise
 */
bool DynamicProgrammingScheduler::hasConflict(const Activity& a1, const Activity& a2) {
    return !(a1.end <= a2.start || a2.end <= a1.start);
}

File: src/algorithms/dynamic_programming.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Dynamic Programming Algorithm for Conflict-Free Scheduling

This module implements dynamic programming for weighted activity selection.
Uses memoization for optimal performance.
"""

from typing import List, Dict
from dataclasses import dataclass


@dataclass
class Activity:
    """Class representing an activity/task with start and end times"""
    id: int
    start: int
    end: int
    weight: float = 1.0  # Default weight for unweighted problems
    name: str = ""
    room: str = ""


class DynamicProgrammingScheduler:
    """Dynamic Programming Algorithm for Weighted Activity Selection"""
    
    @staticmethod
    def solve_weighted_activity_selection(activities: List[Activity]) -> List[Activity]:
        """
        Solve weighted activity selection using dynamic programming
        
        Args:
            activities: List of activities to schedule
            
        Returns:
            List of selected activities with maximum weight
            
        Time Complexity: O(n log n), Space Complexity: O(n)
        """
        if not activities:
            return []
        
        # Sort activities by end time
        activities = sorted(activities, key=lambda a: a.end)
        n = len(activities)
        
        # Build memoization table for dynamic programming
        memo = {}
        
        # Find the latest non-conflicting activity for each activity
        latest_non_conflicting = []
        for i in range(n):
            latest = DynamicProgrammingScheduler.find_latest_non_conflicting(activities, i)
            latest_non_conflicting.append(latest)
        
        def dp(index: int) -> float:
            """
            DP function to find maximum weight up to index
            
            Args:
                index: Current activity index
                
            Returns:
                Maximum weight achievable up to this index
            """
            if index < 0:
                return 0
            
            if index in memo:
                return memo[index]
            
            # Option 1: Don't include current activity
            without_current = dp(index - 1)
            
            # Option 2: Include current activity
            with_current = activities[index].weight
            if latest_non_conflicting[index] >= 0:
                with_current += dp(latest_non_conflicting[index])
            
            memo[index] = max(without_current, with_current)
            return memo[index]
        
        # Calculate maximum weight
        max_weight = dp(n - 1)
        
        # Reconstruct the solution
        selected_activities = []
        i = n - 1
        while i >= 0:
            # Check if current activity is included in optimal solution
            without_current = dp(i - 1) if i > 0 else 0
            with_current = activities[i].weight
            if latest_non_conflicting[i] >= 0:
                with_current += dp(latest_non_conflicting[i])
            
            if with_current >= without_current:
                # Current activity is included
                selected_activities.append(activities[i])
                i = latest_non_conflicting[i]
            else:
                # Current activity is not included
                i -= 1
        
        # Reverse to get chronological order
        selected_activities.reverse()
        
        return selected_activities
    
    @staticmethod
    def find_latest_non_conflicting(activities: List[Activity], index: int) -> int:
        """
        Find the latest non-conflicting activity before the given index
        
        Args:
            activities: List of activities (sorted by end time)
            index: Current activity index
            
        Returns:
            Index of latest non-conflicting activity, or -1 if none found
        """
        for i in range(index - 1, -1, -1):
            if activities[i].end <= activities[index].start:
                return i
        return -1
    
    @staticmethod
    def solve_unweighted_activity_selection(activities: List[Activity]) -> List[Activity]:
        """
        Solve unweighted activity selection (classic greedy approach)
        
        Args:
            activities: List of activities to schedule
            
        Returns:
            List of selected activities (maximum count)
            
        Time Complexity: O(n log n), Space Complexity: O(1)
        """
        if not activities:
            return []
        
        # Sort activities by end time
        activities = sorted(activities, key=lambda a: a.end)
        
        selected = []
        last_end_time = float('-inf')
        
        for activity in activities:
            if activity.start >= last_end_time:
                selected.append(activity)
                last_end_time = activity.end
        
        return selected
    
    @staticmethod
    def solve_with_different_strategies(activities: List[Activity]) -> Dict[str, List[Activity]]:
        """
        Solve using different DP strategies for comparison
        
        Args:
            activities: List of activities to schedule
            
        Returns:
            Dictionary containing results from different strategies
        """
        results = {}
        
        # Strategy 1: Maximum weight
        results['max_weight'] = DynamicProgrammingScheduler.solve_weighted_activity_selection(activities)
        
        # Strategy 2: Maximum count (unweighted)
        results['max_count'] = DynamicProgrammingScheduler.solve_unweighted_activity_selection(activities)
        
        # Strategy 3: Earliest deadline first
        earliest_deadline = sorted(activities, key=lambda a: a.end)
        results['earliest_deadline'] = DynamicProgrammingScheduler._greedy_selection(earliest_deadline)
        
        # Strategy 4: Shortest job first
        shortest_job = sorted(activities, key=lambda a: a.end - a.start)
        results['shortest_job'] = DynamicProgrammingScheduler._greedy_selection(shortest_job)
        
        return results
    
    @staticmethod
    def _greedy_selection(activities: List[Activity]) -> List[Activity]:
        """
        Helper method for greedy selection
        
        Args:
            activities: Pre-sorted list of activities
            
        Returns:
            List of selected activities using greedy approach
        """
        if not activities:
            return []
        
        selected = [activities[0]]
        last_end_time = activities[0].end
        
        for i in range(1, len(activities)):
            if activities[i].start >= last_end_time:
                selected.append(activities[i])
                last_end_time = activities[i].end
        
        return selected
